### PR TITLE
Add `terminal` task

### PR DESCRIPTION
**Pull request by eatkins**
This PR adds a terminal task so that users can access and modify some properties of the terminal that initiated the task. There is some overlap with interactionService but it wasn't clear to me that they should be combined. The `terminal` tasks is a bit more raw. It doesn't provide a `readLine` method or anything like that. Instead it is designed to be just powerful enough for advanced users to build a custom line reader without exposing the entire jline api. The `interactionService` is still probably the way to go for most use cases.

I'd been thinking about adding something like this for a while. The motivation for actually doing it was @eed3si9n's comment in #5899 (https://github.com/sbt/sbt/issues/5899#issuecomment-699467667).

### CODE REVIEW

**Comment by eatkins** [comment]
The main feature that this adds compared to the interactionService that is easy to implement is press any key to exit. Here is a sample build.sbt that does this:
```
TaskKey[Unit]("readInput") := {                                     
  terminal.value.setMode(canonical = false, echo = false)
  terminal.value.printStream.println("read a byte")
  terminal.value.inputStream.read                                   
} 
```

The task exits when the user types any key both with sbt and sbtn.

File: main/src/main/scala/sbt/Terminal.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  /**
+   * Sets the mode of the terminal. By default,the terminal will be in canonical mode
+   * with echo enabled. This means that the terminal's inputStream will not return any
+   * bytes until a newline is received and that all of the characters inputed by the
+   * user will be echoed to the terminal's output stream.
+   *
+   * @param canonical toggles whether or not the terminal input stream is line buffered
+   * @param echo toggles whether or not to echo the characters received from the terminal input stream
+   */
+  def setMode(canonical: Boolean, echo: Boolean): Unit
```
Comment (eed3si9n) [inline main/src/main/scala/sbt/Terminal.scala:61]: Nice documentation here.

**Review by eed3si9n — APPROVED** [review]

**Comment by eatkins** [comment]
I'm sort of wondering now if `Terminal` should be in the sbt package. Is there another package that might make sense that is not internal? The goal was to make this limited functionality public and not in an internal package but top level sbt might be promoting it a little too much. I also wouldn't have had to alias the internal Terminal everywhere it was used in the main project  if it was in a different package.

**Comment by eed3si9n** [comment]
You could make up some subpackage under sbt if you want, but not sure if it's worth it for one class.

**Pull request merged**